api/build/graph.go

```diff
@@ -0,0 +1,576 @@
+
+			return
+		}
+
+		// add page of steps to list steps
+		steps = append(steps, stepsPart...)
+
+		// assume no more pages exist if under 100 results are returned
+		//
+		// nolint: gomnd // ignore magic number
+		if len(stepsPart) < 100 {
+			page = 0
+		} else {
+			page++
+		}
+	}
+
+	if len(steps) == 0 {
+		retErr := fmt.Errorf("no steps found for build %s", entry)
+		util.HandleError(c, http.StatusNotFound, retErr)
+		return
+	}
+
+	// retrieve the services for the build from the service table
+	services := []*library.Service{}
+	page = 1
+	perPage = 100
+	for page > 0 {
+		// retrieve build services (per page) from the database
+		servicesPart, _, err := database.FromContext(c).ListServicesForBuild(ctx, b, nil, page, perPage)
+		if err != nil {
+			retErr := fmt.Errorf("unable to retrieve services for build %s: %w", entry, err)
+
+			util.HandleError(c, http.StatusNotFound, retErr)
+
+			return
+		}
+
+		// add page of services to list services
+		services = append(services, servicesPart...)
+
+		// assume no more pages exist if under 100 results are returned
+		//
+		// nolint: gomnd // ignore magic number
+		if len(servicesPart) < 100 {
+			page = 0
+		} else {
+			page++
+		}
+	}
+
+	logger.Info("retrieving pipeline configuration")
+
+	var config []byte
+
+	lp, err := database.FromContext(c).GetPipelineForRepo(ctx, b.GetCommit(), r)
+	if err != nil { // assume the pipeline doesn't exist in the database yet (before pipeline support was added)
+		// send API call to capture the pipeline configuration file
+		config, err = scm.FromContext(c).ConfigBackoff(ctx, u, r, b.GetCommit())
+		if err != nil {
+			retErr := fmt.Errorf("unable to get pipeline configuration for %s: %w", r.GetFullName(), err)
+
+			util.HandleError(c, http.StatusNotFound, retErr)
+
+			return
+		}
+	} else {
+		config = lp.GetData()
+	}
+
+	if config == nil {
+		retErr := fmt.Errorf("unable to get pipeline configuration for %s: config is nil", r.GetFullName())
+
+		util.HandleError(c, http.StatusNotFound, retErr)
+
+		return
+	}
+
+	// variable to store changeset files
+	var files []string
+	// check if the build event is not issue_comment
+	if !strings.EqualFold(b.GetEvent(), constants.EventComment) {
+		// check if the build event is not pull_request
+		if !strings.EqualFold(b.GetEvent(), constants.EventPull) {
+			// send API call to capture list of files changed for the commit
+			files, err = scm.FromContext(c).Changeset(ctx, u, r, b.GetCommit())
+			if err != nil {
+				retErr := fmt.Errorf("%s: failed to get changeset for %s: %v", baseErr, r.GetFullName(), err)
+
+				util.HandleError(c, http.StatusInternalServerError, retErr)
+
+				return
+			}
+		}
+	}
+
+	logger.Info("compiling pipeline configuration")
+
+	// parse and compile the pipeline configuration file
+	p, _, err := compiler.FromContext(c).
+		Duplicate().
+		WithBuild(b).
+		WithFiles(files).
+		WithMetadata(m).
+		WithRepo(r).
+		WithUser(u).
+		Compile(config)
```

Comment:
```suggestion
	// parse and compile the pipeline configuration file
	p, _, err := compiler.FromContext(c).
		Duplicate().
		WithBuild(b).
		WithFiles(files).
		WithCommit(b.GetCommit())
		WithMetadata(m).
		WithRepo(r).
		WithUser(u).
		Compile(config)
```

Will be necessary for `file` type templates.

Ref: https://github.com/go-vela/server/blob/4eec5e9b03ab7e7004bfae08fc558eb2b27d87c5/compiler/native/expand.go#L289

Reply:
oo updated, thank you